### ISSUE TITLE
Add a feature test before enabling _FILE_OFFSET_BITS macro

### ISSUE THREAD

**Issue by siteshwar** [mesonbuild/meson#3519]
Copying my comment from #1455.

Support for 64-bit operations in fts functions was added only a few years ago in glibc [through this commit](https://sourceware.org/git/?p=glibc.git;a=commit;h=8b7b7f75d91f7bac323dd6a370aeb3e9c5c4a7d5) and there are systems that still ship older versions glibc (for e.g. opensuse). Unconditionally enabling this macro on all systems will cause compilation failures. There should be a feature test before this macro is enabled.

**Comment by siteshwar** [comment]
For the reference, test case should be pretty simple:
```
> cat test.c
#include <fts.h>

int main() {
}

> cc -D_FILE_OFFSET_BITS=64 test.c
In file included from test.c:1:0:
/usr/include/fts.h:41:3: error: #error "<fts.h> cannot be used with -D_FILE_OFFSET_BITS==64"
 # error "<fts.h> cannot be used with -D_FILE_OFFSET_BITS==64"
#include <fts.h>

int main() {
}
```

**Comment by nirbheek** [comment]
See also: https://github.com/mesonbuild/meson/issues/3049. Unconditionally enabling this is broken on some platforms (Raspbian armhf) and hilariously broken on other platforms (Android 32-bit).

We do want to fix this properly, but no one has come along who is willing to investigate the edge cases involved. Would you like to work on this?

**Comment by siteshwar** [comment]
@nirbheek Is there a way to turn off this flag in builds ? If I am reading the docs correctly, it can not be turned off, so it should not be added to builds by default. I might look at it in my free time, but can not promise a fix at the moment.

**Comment by nirbheek** [comment]
> @nirbheek Is there a way to turn off this flag in builds ?

Yes, if you pass `-U_FILE_OFFSET_BITS` in `add_project_arguments()` (or on the specific target with `c_args:`) that should do it.

**Comment by siteshwar** [comment]
@nirbheek Thanks for the quick fix.

**Comment by nirbheek** [comment]
Notes: 

1. This is a bug in glibc that we have to work around
2. The Android NDK bug was fixed when targeting Android SDK >=15
3. The raspbian armhf bug has not been reproduced yet

**Comment by mid-kid** [comment]
Bumping this, it bit my ass when converting a program from a different build system to meson. I think there should be a toggle at least - no other build system enforces it like that.
